src/Module/BaseApi.php

```diff
@@ -178,97 +170,115 @@ public static function getRequest(array $defaults)
 	}
 
 	/**
-	 * Log in user via OAuth1 or Simple HTTP Auth.
-	 *
-	 * Simple Auth allow username in form of <pre>user@server</pre>, ignoring server part
+	 * Check the logged in user is allowed to use the provided API scope. Aborts request if not logged or
+	 * insufficient scopes.
 	 *
 	 * @param string $scope the requested scope (read, write, follow)
 	 *
-	 * @return bool Was a user authenticated?
 	 * @throws HTTPException\ForbiddenException
 	 * @throws HTTPException\UnauthorizedException
 	 * @throws HTTPException\InternalServerErrorException
-	 * @hook  'authenticate'
-	 *               array $addon_auth
-	 *               'username' => username from login form
-	 *               'password' => password from login form
-	 *               'authenticated' => return status,
-	 *               'user_record' => return authenticated user record
 	 */
-	protected static function login(string $scope)
+	protected static function checkAllowedScope(string $scope)
```

Comment:
I find the new name slightly irritating, since it's first task is to log in a user. And only on OAuth but not basic auth it does check the scope as well.

Reply:
I agree with you, currently it is doing a lot of things at the same time. Ideally, we should pass the token to this method and it would check the allowed scopes from the token. Now, the method that retrieves the token should be passed the HTTP header(s) to be able to retrieve that information explicitly, and so on. This would be the right way to do it, provide each method with the explicit data they need to perform their functionality, instead of relying on superglobal variables.

But we aren't there yet in terms of decoupling, and it would require us to add a lot of boilerplate code to the existing API calls which probably isn't in the scope of this PR. And since `checkAllowedScope` doesn't perform login if it was already performed before in the code, then it didn't feel right to me to name it as such.

Reply:
I think that "login" is some kind of signal word. It tells that you need to be logged in to continue. So on a first glance you already see that this function requires a login. I want to avoid any possibility of security problems. Another solution would be to have a single public function that fetches the user id and is called `getUserIdForScope()` that would need to be called with a scope or with `self::SCOPE_PUBLIC` if anonymous access was allowed. If there is only one function to fetch the user id and we have to provide the scope with it, it should lower the danger of accidentally provide public access where it shouldn't. I always have to think of the problem that we had got with the user export lately.

Reply:
The problem is that `login` is way too generic a word. It doesn’t say anything about what is actually performed nor what the expected outcome/return is. The name `checkAllowedScope` carries more information in itself:
- `check`: there’s no expected return, it’s just an assertion.
- `allowed`: Now we’re touching access control, which may include authentication.
- `scope`: This is where we provide the API endpoint specific requirement.

`getUserIdForScope` is similarly meaningful, but it doesn’t adequately cover anonymous endpoint that don’t need a user id nor a scope check to work as expected.

Reply:
I really would like to have the scope check and the fetching of the user id combined in a single function. It is easy to forget this check when its done in separate functions. But this can then create security problems, because some activities are possible with an empty user id as well. But by having the need to provide a scope when fetching the user id, one cannot forget it.

Reply:
If defining a scope is required for authenticated API endpoints, then the name `getUserIdForAccessScope` is a good way to combine both checks and return a value the endpoint meeds.

It just won’t be called in anonymous endpoints.

Reply:
There are endpoints that can be called with an user account but needn't. Wit an user they possibly just reveal more data. So every call for fetching a user id can have a scope. Only thing is: For the public access there would be a new scope. And the best thing is: during the coding one has to decide which scope to chose.

Reply:
Then if these concepts are independent, we should keep using two different methods for scope checking and user id retrieval.

Reply:
From a security point of view we should combine this. Otherwise the possibility is high that at one point we forget the scope check.

Reply:
I don’t the see the difference in potential security liability between having a separate method to check scope compared to having a `SCOPE_PUBLIC` constant which use that could be wrongly copied and pasted from another endpoint.

Reply:
The human brain recognizes wrong data much better than missing data. So a wrong scope is much more easily detectable than a missing function call.

Reply:
I am not convinced. From experience in this very project, clumping features together seems to be more error-prone.

Reply:
Be prepared that I will point you to this discussion if the described situation should happen in the future 😁